### PR TITLE
Remove foreign key from the wc_download_log table

### LINKED ISSUE #34305: [Enhancement]: Remove foreign key for download permissions
### Describe the solution you'd like

Replace the FK constraint with a code solution.

### Describe alternatives you've considered

n/a

### Additional context

We don't use FKs for modeling any other relationships in our DB schema and since using a FK forces the use of InnoDB, it's not really necessary even here. 
It was introduced as part of GDPR readiness (PR https://github.com/woocommerce/woocommerce/pull/19330) and doesn't seem necessary.

### PR DESCRIPTION
**Pull request by peterfabian**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #34305  .

This is the only use of FK in our whole db schema and seems unnecessary, as it forces the use of InnoDB engine. We don't use FKs elsewhere, so it's better to remove it here as well.

### How to test the changes in this Pull Request:

1.  Run the upgrade and make sure the foreign key was removed.
2. Create a downloadable product
3. Buy it a couple of times with different customers and download the file(s) to create download log entries.
4. Check the permission table (`wp_woocommerce_downloadable_product_permissions`) and download log table (`wp_wc_download_log`) to make sure there are some records 
5. Try to programmatically remove the download log through all the data stores/abstractions above.
6. Set some permissions as expired (change the date to yesterday) in the database (e.g. via phpMyAdmin)
7. Go to WC > Settings > Tools and run the tool that removes expired access rights to files.
8. Make sure they were removed.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by peterfabian** [comment]
Rebased on top of https://github.com/woocommerce/woocommerce/pull/34416 to include the tests.

File: plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php [hunk 1]
```diff
@@ -240,11 +240,20 @@ public function update( &$download ) {
...
+			)
+		);
+		// Delete related records in wc_download_log (aka ON DELETE CASCADE).
+		$wpdb->query(
+			$wpdb->prepare(
+				"DELETE FROM {$wpdb->prefix}wc_download_log
+				WHERE permission_id = %d",
+				$download_id
 			)
 		);
```
Comment (Konamiman) [inline plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php]: This code is repeated verbatim shortly after. Could it be refactored into a `delete_download_log_by_permission_id` method (private and in the same class)?

File: plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php [hunk 2]
```diff
@@ -274,6 +291,18 @@ public function delete_by_id( $id ) {
...
+			$wpdb->prepare(
+				"DELETE FROM {$wpdb->prefix}wc_download_log
+				WHERE permission_id IN (
+				    SELECT permission_id
+				    FROM {$wpdb->prefix}woocommerce_downloadable_product_permissions
+				    WHERE order_id = %d
+				)",
+				$id
+			)
+		);
```
Comment (Konamiman) [inline plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php]: Similarly, this code is repeated four times, changing only the field of the `where`. I think a `delete_download_log_by_field_value($field, $value)` method would be appropriate here.

File: plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php (hunk 1 above)
Comment (peterfabian) [reply plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php]: Good point. I think I can just call `$this->delete_by_id( $download_id )` and that will take care of both queries.

Comment (peterfabian) [reply plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php]: Should be fixed in[461d5c4](https://github.com/woocommerce/woocommerce/pull/34318/commits/461d5c4bb6d50cf752242b00a4d0ca70b5dcc227).

File: plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php (hunk 2 above)
Comment (peterfabian) [reply plugins/woocommerce/includes/data-stores/class-wc-customer-download-data-store.php]: True. Extracted to a private method in [461d5c4](https://github.com/woocommerce/woocommerce/pull/34318/commits/461d5c4bb6d50cf752242b00a4d0ca70b5dcc227).

**Review by Konamiman — APPROVED** [review]

**Pull request merged**